core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -416,6 +417,43 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
         LUCENE_DOC_NOT_FOUND
     }
 
+    private OpVsLuceneDocStatus compareOpToLuceneDocBasedOnSeqNo(final Operation op) throws IOException {
+        assert op.seqNo() != SequenceNumbersService.UNASSIGNED_SEQ_NO : "resolving ops based seq# but no seqNo is found";
+        final OpVsLuceneDocStatus status;
+        final VersionValue versionValue = versionMap.getUnderLock(op.uid());
+        assert incrementVersionLookup();
+        if (versionValue != null) {
+            if  (op.seqNo() > versionValue.getSeqNo() ||
+                (op.seqNo() == versionValue.getSeqNo() && op.primaryTerm() > versionValue.getTerm()))
+                status = OpVsLuceneDocStatus.OP_NEWER;
```

Comment:
in general I find it odd that we have `OP_STALE_OR_EQUAL` instead of `OP_NEWER_OR_EQUAL` which is more natural then you have only one state where indexing would be fatal but I can see that this is an optimization to not index the doc again?